test/upgrade/continual/source.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package continual
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/google/uuid"
+	corev1 "k8s.io/api/core/v1"
+	"knative.dev/eventing-kafka/test/e2e/helpers"
+	contribtestlib "knative.dev/eventing-kafka/test/lib"
+	contribresources "knative.dev/eventing-kafka/test/lib/resources"
+	"knative.dev/eventing/test/upgrade/prober"
+	"knative.dev/eventing/test/upgrade/prober/sut"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+	pkgTest "knative.dev/pkg/test"
+	pkgupgrade "knative.dev/pkg/test/upgrade"
+)
+
+const (
+	defaultKafkaBootstrapPort    = 9092
+	defaultKafkaClusterName      = "my-cluster"
+	defaultKafkaClusterNamespace = "kafka"
+	sourceConfigTemplatePath     = "test/upgrade/continual/source-config.toml"
+)
+
+// SourceTest tests source operation in continual manner during the
+// whole upgrade and downgrade process asserting that all event are propagated
+// well.
+func SourceTest(opts *TestOptions) pkgupgrade.BackgroundOperation {
+	return continualVerification(
+		"SourceContinualTest",
+		ensureKafkaSender(opts),
+		&kafkaSourceSut{KafkaCluster: opts.KafkaCluster},
+		sourceConfigTemplatePath,
+	)
+}
+
+func ensureKafkaSender(opts *TestOptions) *TestOptions {
+	opts.Configurators = append([]prober.Configurator{
+		func(config *prober.Config) error {
+			config.Wathola.ImageResolver = kafkaSourceSenderImageResolver
+			return nil
+		},
+	}, opts.Configurators...)
+	return opts
+}
+
+func kafkaSourceSenderImageResolver(component string) string {
+	if component == "wathola-sender" {
+		// replacing the original image with modified one from this repo
+		component = "wathola-kafka-sender"
+	}
+	return pkgTest.ImagePath(component)
+}
+
+type kafkaSourceSut struct {
+	KafkaCluster
+}
+
+func (k kafkaSourceSut) Deploy(ctx sut.Context, destination duckv1.Destination) interface{} {
+	topicName := uuid.NewString()
+	c := k.KafkaCluster
+	helpers.MustCreateTopic(ctx.Client, c.name(), c.namespace(),
+		topicName, 6)
+	contribtestlib.CreateKafkaSourceV1Beta1OrFail(ctx.Client, contribresources.KafkaSourceV1Beta1(
+		c.serversLine(),
+		topicName,
+		toObjectReference(destination),
+	))
+	return kafkaTopicEndpoint{
+		BootstrapServers: k.KafkaCluster.serversLine(),
+		TopicName:        topicName,
+	}
+}
+
+func (c KafkaCluster) name() string {
+	if c.Name == "" {
```

Comment:
It seems like it would have been better to fill in the defaults when creating the kafka cluster via `fillInDefaults` and in this place only return `c.Name`` instead of returning some default value which might not be correct. Just shifting the logic a little bit.

Reply:
Hmm, maybe. I'll take a look...

Reply:
I've made a refactor on how to pass those variables and set default values. Now it's also possible to influence retry and replication options.

Changes are in this commit: https://github.com/knative-sandbox/eventing-kafka/pull/563/commits/5fa1ec780d22cc4b4416db84727b87250cfce709

Reply:
Nice!